lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,341 @@
+          std::get<2>(*selectCaseIntRangeValsIter), source);
+    }
+  }
+  void Insert(std::optional<std::string> lowerVal,
+      std::optional<std::string> upperVal, parser::CharBlock source) {
+    const auto &stringRangeValue{std::make_tuple(lowerVal, upperVal, source)};
+    selectCaseValsIter = std::find_if(selectCaseVals.begin(),
+        selectCaseVals.end(), [&stringRangeValue](const auto &mem) {
+          return (
+              (!std::get<0>(stringRangeValue) ||
+                  ((std::get<0>(stringRangeValue).value())
+                          .compare(std::get<std::string>(mem.first)) >= 0)) &&
+              (!std::get<1>(stringRangeValue) ||
+                  ((std::get<1>(stringRangeValue).value())
+                          .compare(std::get<std::string>(mem.first)) <= 0)));
+        });
+    if (selectCaseValsIter != selectCaseVals.end()) {
+      ReportOverlapCaseValError(selectCaseValsIter->second, source);
+    }
+
+    selectCaseStringRangeValsIter = std::find_if(
+        selectCaseStringRangeVals.begin(), selectCaseStringRangeVals.end(),
+        [&stringRangeValue](const auto &mem) {
+          return ((!std::get<0>(mem) && !std::get<0>(stringRangeValue)) ||
+              (!std::get<1>(mem) && !std::get<1>(stringRangeValue)) ||
+              ((!std::get<0>(mem) ||
+                   (std::get<0>(stringRangeValue) &&
+                       ((std::get<0>(stringRangeValue).value())
+                               .compare(std::get<0>(mem).value()) >= 0))) &&
+                  (!std::get<1>(mem) ||
+                      (std::get<0>(stringRangeValue) &&
+                          ((std::get<0>(stringRangeValue).value())
+                                  .compare(std::get<1>(mem).value()) <= 0)))) ||
+              ((!std::get<0>(mem) ||
+                   (std::get<1>(stringRangeValue) &&
+                       (std::get<1>(stringRangeValue)
+                               .value()
+                               .compare(std::get<0>(mem).value()) >= 0))) &&
+                  (!std::get<1>(mem) ||
+                      (std::get<1>(stringRangeValue) &&
+                          (std::get<1>(stringRangeValue)
+                                  .value()
+                                  .compare(std::get<1>(mem).value()) <= 0)))));
+        });
+
+    if (selectCaseStringRangeValsIter == selectCaseStringRangeVals.end()) {
+      selectCaseStringRangeVals.push_back(stringRangeValue);
+    } else {
+      ReportOverlapCaseValError(
+          std::get<2>(*selectCaseStringRangeValsIter), source);
+    }
+  }
+
+private:
+  using selectCaseStmtTypes =
+      std::vector<std::pair<std::variant<std::int64_t, bool, std::string>,
+          parser::CharBlock>>;
+  selectCaseStmtTypes selectCaseVals;
+  selectCaseStmtTypes::iterator selectCaseValsIter;
+  using selectCaseIntRangeType =
+      std::vector<std::tuple<std::optional<std::int64_t>,
+          std::optional<std::int64_t>, parser::CharBlock>>;
```

Comment:
If you defined a struct for the members of the vector the data members would have meaningful names and you wouldn't have to use std::get<0>, etc.

Reply:
Since only first two members of tuples are being used frequently, I thought it is a overhead to create a struct and handle instead of using tuple. Should I modify them and use struct instead?

Reply:
Use a struct or class.

Reply:
Ok. I will replace tuple with a struct.